chain/vm/syscalls.go

```diff
@@ -46,15 +57,141 @@ func (ss *syscallShim) ComputeUnsealedSectorCID(st abi.RegisteredProof, pieces [
+// Note that the blocks are ordered: the method requires a.Epoch() <= b.Epoch().
 func (ss *syscallShim) VerifyConsensusFault(a, b, extra []byte, epoch abi.ChainEpoch) (*runtime.ConsensusFault, error) {
-	panic("NYI")
+	// Note that block syntax is not validated. Any validly signed block will be accepted pursuant to the below conditions.
+	// Whether or not it could ever have been accepted in a chain is not checked/does not matter here.
+	// for that reason when checking block parent relationships, rather than instantiating a Tipset to do so
+	// (which runs a syntactic check), we do it directly on the CIDs.
+
+	// (0) cheap preliminary checks
+
+	// are blocks the same?
+	if bytes.Equal(a, b) {
+		return nil, fmt.Errorf("no consensus fault: submitted blocks are the same")
+	}
+
+	// can blocks be decoded properly?
+	var blockA, blockB types.BlockHeader
+	if decodeErr := blockA.UnmarshalCBOR(bytes.NewReader(a)); decodeErr != nil {
+		return nil, xerrors.Errorf("cannot decode first block header: %w", decodeErr)
+	}
+
+	if decodeErr := blockB.UnmarshalCBOR(bytes.NewReader(b)); decodeErr != nil {
+		return nil, xerrors.Errorf("cannot decode second block header: %f", decodeErr)
+	}
+
+	// (1) check conditions necessary to any consensus fault
+
+	// were blocks mined by same miner?
+	if blockA.Miner != blockB.Miner {
+		return nil, fmt.Errorf("no consensus fault: blocks not mined by same miner")
+	}
+
+	// block a must be earlier or equal to block b, epoch wise (ie at least as early in the chain).
+	if blockB.Height < blockA.Height {
+		return nil, fmt.Errorf("first block must not be of higher height than second")
+	}
+
+	// (2) check for the consensus faults themselves
+	var consensusFault *runtime.ConsensusFault
+
+	// (a) double-fork mining fault
+	if blockA.Height == blockB.Height {
+		consensusFault = &runtime.ConsensusFault{
+			Target: blockA.Miner,
+			Epoch:  blockB.Height,
+			Type:   runtime.ConsensusFaultDoubleForkMining,
+		}
+	}
+
+	// (b) time-offset mining fault
+	// strictly speaking no need to compare heights based on double fork mining check above,
+	// but at same height this would be a different fault.
+	if !types.CidArrsEqual(blockA.Parents, blockB.Parents) && blockA.Height != blockB.Height {
+		consensusFault = &runtime.ConsensusFault{
+			Target: blockA.Miner,
+			Epoch:  blockB.Height,
+			Type:   runtime.ConsensusFaultTimeOffsetMining,
+		}
+	}
+
+	// (c) parent-grinding fault
+	// Here extra is the "witness", a third block that shows the connection between A and B as
+	// A's sibling and B's parent.
+	// Specifically, since A is of lower height, it must be that B was mined omitting A from its tipset
+	var blockC types.BlockHeader
+	if len(extra) > 0 {
+		if decodeErr := blockC.UnmarshalCBOR(bytes.NewReader(extra)); decodeErr != nil {
+			return nil, xerrors.Errorf("cannot decode extra: %w", decodeErr)
+		}
+
+		if types.CidArrsEqual(blockA.Parents, blockC.Parents) && blockA.Height == blockB.Height &&
```

Comment:
from what I'm reading, `a.Height` and `b.Height` will be always different in this case, code says something else.

also with that check it would be `(a)` case.

Reply:
ah typo, it is `blockA.Height == blockC.Height`. Great catch.

Do you see how it's different from a?

Reply:
yeah, not it makes sense.